### PR TITLE
[FRONTEND][TENSORFLOW] Enhancements.

### PR DESCRIPTION
**Pull request by srkreddy1238**
* Generalize the shape with explicite argument.
	* Supported entire range of mobilenet_v2 models.
	* Cast op updated to latest tensorflow.
	* Documentation updates.
	* CheckNumerics op handling without exception.
	* API to fetch test data from tensorflow official releases.

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

**Comment by srkreddy1238** [comment]
cc @masahi @sgrechanik-h @Huyuwei @FrozenGene welcome to review

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 1]
```diff
@@ -1131,21 +1136,33 @@ def from_tensorflow(self, graph, layout="NHWC"):
...
+                    op = _sym.Variable(name=node.name,
+                                       shape=self._params[node.name].shape)
+
                 # Assuming only one output.
                 self._nodes[node.name] = op
-                node_output = op
+
+            # Infer shapes if passed explicitely
+            node_output = self._nodes[node.name]
+            if shape:
```
Comment (sgrechanik-h) [inline nnvm/python/nnvm/frontend/tensorflow.py]: `shape={}` will behave the same way as `shape=None`. Is it intended? I'm thinking about some hypothetical corner cases like when there are no inputs.

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 2]
```diff
@@ -1091,6 +1094,9 @@ def from_tensorflow(self, graph, layout="NHWC"):
                 self._output_shapes[node.name] = \
                     [tensor_util.TensorShapeProtoToList(shape) \
                     for shape in attr['_output_shapes']]
+            elif shape:
+                # Keep some value to avoid key error and infer shape after node creation.
+                self._output_shapes[node.name] = [[1,]]
```
Comment (sgrechanik-h) [inline nnvm/python/nnvm/frontend/tensorflow.py]: Why `[[1]]` and not `None` or `[[]]`?

Comment (srkreddy1238) [reply nnvm/python/nnvm/frontend/tensorflow.py]: ```[None]``` or ```[[ ]]``` should work. I will change.

**Comment by srkreddy1238** [comment]
@sgrechanik-h  please have a look.
I have tested all models (MobilenetV1/V2, Resnet, Inception, VGG) from research/slim.

**Comment by srkreddy1238** [comment]
@tqchen 
I have made some utils and documentation related to generating model protobuf files from tensorflow official checkpoints at https://github.com/srkreddy1238/dmlc_data/tree/master/work/tf/samples.

Do you think we could accommodate them in TVM as docs & utils?
Or 
Make them as test cases but deactivate?

**Comment by tqchen** [comment]
In the interest of minimum codebase, let us keep it as a separate thing, but we can generate the proto and download them in the testcase. Thanks @srkreddy1238

**Comment by srkreddy1238** [comment]
@tqchen 
I am not talking about adding the protobuf files into TVM.
Also tensorflow research/slim models are nearly 40, adding all of them into test cases delays the CI time.
Some of them need building few utils from tensorflow source too (Resnet, Vgg) to prepare protobuf from checkpoint.
My interest here is about sharing the steps(tutorials) from tensorflow release to NNVM frontend import.

What do you suggest ?

**Comment by tqchen** [comment]
Maybe we can build up another set of test that runs nightly, and the normal fast test jobs to make sure general coverage

**Comment by tqchen** [comment]
Another possible approach is to provide the conversion dry run, which converts the model but does not run them. This can considerably reduce the test time and can be placed in the CI

**Comment by srkreddy1238** [comment]
+1 for nightly build.

Conversion dry run is not a straight forward as checkpoint to protobuf conversion need https://github.com/tensorflow/models (to generate initial protobuf) and freeze_graph built from tensorflow source to embed checkpoint into it.

I may revisit into simplifying/automating this later for nightly build. For now I will leave a doc @ docs/frontend/tensorflow.md.

**Comment by srkreddy1238** [comment]
@nishi-t  welcome to review.

File: docs/frontend/tensorflow.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+This document helps few steps while importing various different models from
+[tensorflow research/slim](https://github.com/tensorflow/models/tree/master/research/slim).
+
+Current frontend is tested with all versions of below models
+- Inception (V1/V2/V3/V4)
+- Resnet (All)
+- Mobilenet (V1/V2 All)
+- Vgg (16/19)
+
+Tensorflow frontend expects a feeezed protobuf format as input.
```
Comment (siju-samuel) [inline docs/frontend/tensorflow.md]: freezed

File: docs/frontend/tensorflow.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+Not all models are released as freezed protobuf. Soem of tyhem are checkpoints (.ckpt).
```
Comment (siju-samuel) [inline docs/frontend/tensorflow.md]: Soem of tyhem -> Some of them

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 5]
```diff
@@ -1030,28 +1039,33 @@ def __init__(self):
...
-
         Parameters
         ----------
         graph : tensorflow graph definition object
             The loaded tensorflow GraphDef
 
+        layout : target layout to be used (Optional)
+            NCHW only supported now to enable NHWC models on GPU.
+
+        shape : Dictionary of input dimentions (Optional)
```
Comment (siju-samuel) [inline nnvm/python/nnvm/frontend/tensorflow.py]: dimensions

File: nnvm/python/nnvm/testing/tf.py [hunk 6]
```diff
@@ -135,13 +136,19 @@ def id_to_string(self, node_id):
             return ''
         return self.node_lookup[node_id]
 
-def get_workload(model_path):
-    """ Import workload from frozen protobuf
+def get_workload_official(model_url, model_sub_path, temp):
+    """ Import workload from tensorflow official
```
Comment (siju-samuel) [inline nnvm/python/nnvm/testing/tf.py]: instead of `temp`, use `directory` or `temp_dir`

File: nnvm/python/nnvm/testing/tf.py [hunk 7]
```diff
@@ -150,16 +157,49 @@ def get_workload(model_path):
...
+
+    download(model_url, path_model)
+
+    import tarfile
+    if path_model.endswith("tgz") or path_model.endswith("gz"):
+        tar = tarfile.open(path_model)
+        tar.extractall(path=temp_path)
+        tar.close()
+    else:
+        raise RuntimeError('Could not un compress the file: ' + path_model)
```
Comment (siju-samuel) [inline nnvm/python/nnvm/testing/tf.py]: un compress -> uncompress/decompress

File: nnvm/python/nnvm/testing/tf.py [hunk 8]
```diff
@@ -150,16 +157,49 @@ def get_workload(model_path):
...
+    return temp_path + model_sub_path
+
+def get_workload(model_path, model_sub_path=None):
+    """ Import workload from frozen protobuf
+
+    Parameters
+    ----------
+    model_path: str
```
Comment (siju-samuel) [inline nnvm/python/nnvm/testing/tf.py]: docstring dont have model_sub_path

File: nnvm/python/nnvm/frontend/tensorflow.py [hunk 9]
```diff
@@ -1131,21 +1147,33 @@ def from_tensorflow(self, graph, layout="NHWC"):
                     node.input[0] = in_name
 
                 # Fill shapes for all inputs in a list
-                try:
-                    inputs = [self._nodes[i] for i in node.input]
-                    for i in node.input:
+                inputs = []
+                for i in node.input:
+                    if i in self._nodes:
+                        inputs.append(self._nodes[i])
```
Comment (nishi-t) [inline nnvm/python/nnvm/frontend/tensorflow.py]: I found the case that causes KeyError in `self._nodes[i]` in [my current work.](#2001 ) I think that it is needed to handle `KeyError`. I'd appreciate your help.

For example: 
```
    def check_split_concat(ishape, **kwargs):
        inp_array = np.random.uniform(size=ishape).astype(np.float32)
        with tf.Graph().as_default():
            in1 = tf.placeholder(shape=inp_array.shape, dtype=inp_array.dtype)
            splited = tf.split(in1, **kwargs)
            tf.concat(splited, axis=1)
            compare_tf_with_tvm(inp_array, 'Placeholder:0', 'concat:0')

    check_split_concat((5, 30), num_or_size_splits=[15, 15], axis=1)
```
In above code, the concat node has three inputs(`split`, `split1`, `concat/axis`), but `split1` is not registered in _nodes.(the detail of graphdef in this case is [here](dabff8cb4b54302ccc8d269e49baea10)) Moreover, these `split:X` is increased depending on the number of split.

File: docs/frontend/tensorflow.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+## General Instructions
+
+### Add Shapes:
+While freezing of protobuf add additional option ```add_shapes=True``` to embed output shapes of each node into graph.
+You may use ```nnvm.testing.tf.AddShapesToGraphDef``` from nnvm for the same.
+Please refer to [tensorflow tutorial](https://github.com/dmlc/tvm/blob/master/tutorials/nnvm/from_tensorflow.py).
+
+### Explicit Shape:
+There might be situations where the add_shape=True may not provide sufficient information about shape.
```
Comment (nishi-t) [inline docs/frontend/tensorflow.md]: add_shape => add_shapes

File: nnvm/python/nnvm/frontend/tensorflow.py (hunk 9 above)
Comment (nishi-t) [reply nnvm/python/nnvm/frontend/tensorflow.py]: Please don't care above my comment. I understand it is a problem which should be resolved separately from this PR. Thanks ;)

**Comment by srkreddy1238** [comment]
cc @nishi-t have another look pls.

**Comment by yzhliu** [comment]
Thanks! this is now merged.

**Pull request merged**